addon/utils/html-input-parser.ts

```diff
@@ -0,0 +1,155 @@
+
+interface HTMLInputParserArguments {
+  safeAttributes?: string[],
+  lumpTags?: string[],
+  safeTags?: string[],
+  uriSafeAttributes?: string[],
+  tagMap?: Map<string, string>
+}
+
+/**
+ * An html input parser for the editor.
+ * The parser makes the HTML input safe for usage in the editor.
+ * This means it removes any tags, attributes and styling we don't understand.
+ * It may also translate attributes and tags to things we do understand.
+ *
+ * @module ember-rdfa-editor
+ * @class HTMLInputParser
+ */
+export default class HTMLInputParser {
+  static DEFAULTS = {
+    safeAttributes: DEFAULT_SAFE_ATTRIBUTES,
+    lumpTags: DEFAULT_LUMP_TAGS,
+    safeTags: DEFAULT_SAFE_TAGS,
+    uriSafeAttributes: DEFAULT_URI_SAFE_ATTRIBUTES,
+    tagMap: DEFAULT_TAG_MAP
+  };
+
+  private readonly safeAttributes: string[];
+  private readonly lumpTags: string[];
+  private readonly safeTags: string[];
+  private readonly uriSafeAttributes: string[];
+  private readonly tagMap: Map<string, string>;
+
+  constructor({safeAttributes, lumpTags, safeTags, uriSafeAttributes, tagMap}: HTMLInputParserArguments) {
+    this.safeAttributes = safeAttributes ? safeAttributes : DEFAULT_SAFE_ATTRIBUTES;
+    this.lumpTags = lumpTags ? lumpTags : DEFAULT_LUMP_TAGS;
+    this.safeTags = safeTags ? safeTags : DEFAULT_SAFE_TAGS;
+    this.uriSafeAttributes = uriSafeAttributes ? uriSafeAttributes : DEFAULT_URI_SAFE_ATTRIBUTES;
+    this.tagMap = tagMap ? tagMap : DEFAULT_TAG_MAP;
+  }
+
+  /**
+   * Takes an html string, preprocesses its nodes and sanitizes the result.
+   * Returns the cleaned html string.
+   *
+   * @method cleanupHTML
+   */
+  cleanupHTML(htmlString: string): string {
+    const parser = new DOMParser();
+    const document = parser.parseFromString(htmlString, "text/html");
+    const rootNode = document.body;
+    rootNode.normalize();
+
+    const preprocessedNode = this.preprocessNode(rootNode) as HTMLElement; // Assumption: root node can never be text
+    return DOMPurify.sanitize(preprocessedNode.innerHTML, {
+      ALLOWED_TAGS: this.safeTags,
+      ALLOWED_ATTR: this.safeAttributes,
+      // ADD_URI_SAFE_ATTR: this.uriSafeAttributes TODO: does this work?
+    });
+  }
+
+  /**
+   * Preprocess all nodes replacing the tag if it appears on the tagMap variable
+   * and adds the lumpNode property if needed.
+   * TODO: check if we can get rid of casts
+   *
+   * @method preprocessNode
+   */
+  preprocessNode(node: Node) {
+    let cleanedNode = node.cloneNode();
+    if (node.nodeType === Node.ELEMENT_NODE) {
+      const tag = tagName(node);
+      const tagMapping = this.tagMap.get(tag);
+
+      // If we have to replace the tag name we create another node with the new
+      // tag name and copy all the attribute of the original node.
+      if (tagMapping) {
+        cleanedNode = document.createElement(tagMapping);
+        this.copyAllAttributes((node as HTMLElement), (cleanedNode as HTMLElement));
```

Comment:
the reason you need to cast cleanedNode (even though you are sure it's an element here) is that the cleanedNode variable was defined as having the Node type. Probably best to just create a new var here and only overwrite cleanedNode at the end